src/pint/utils.py

```diff
@@ -2764,3 +2741,234 @@ def woodbury_dot(Ndiag, U, Phidiag, x, y):
+
+    hess = Hessian(mlnlike)
+
+    H = hess((gamma_val, log10_A_val))
+    assert np.all(np.linalg.eigvals(H) > 0), "The Hessian is not positive definite!"
+
+    Hinv = np.linalg.pinv(H)
+    assert np.all(
+        np.linalg.eigvals(Hinv) > 0
+    ), "The inverse Hessian is not positive definite!"
+
+    gamma_err, log10_A_err = np.sqrt(np.diag(Hinv))
+
+    tndmc = len(model.components["DMWaveX"].get_indices())
+
+    model1 = deepcopy(model)
+    model1.remove_component("DMWaveX")
+    model1.add_component(PLDMNoise())
+    model1.TNDMAMP.value = log10_A_val
+    model1.TNDMGAM.value = gamma_val
+    model1.TNDMC.value = tndmc
+    model1.TNDMAMP.uncertainty_value = log10_A_err
+    model1.TNDMGAM.uncertainty_value = gamma_err
+
+    return model1
+
+
+def find_optimal_nharms(model, toas, component, nharms_max=45):
+    """Find the optimal number of harmonics for `WaveX`/`DMWaveX` using the Akaike Information
+    Criterion.
+
+    Parameters
+    ----------
+    model: `pint.models.timing_model.TimingModel`
+        The timing model. Should not already contain `WaveX`/`DMWaveX` or `PLRedNoise`/`PLDMNoise`.
+    toas: `pint.toa.TOAs`
+        Input TOAs
+    component: str
+        Component name; "WaveX" or "DMWaveX"
+    nharms_max: int
+        Maximum number of harmonics
+
+    Returns
+    -------
+    nharms_opt: int
+        Optimal number of harmonics
+    aics: ndarray
+        Array of normalized AIC values.
+    """
+    from pint.fitter import Fitter
+
+    assert component in ["WaveX", "DMWaveX"]
+    assert (
+        component not in model.components
+    ), f"{component} is already included in the model."
+    assert (
+        "PLRedNoise" not in model.components and "PLDMNoise" not in model.components
+    ), "PLRedNoise/PLDMNoise cannot be included in the model."
+
+    model1 = deepcopy(model)
+
+    ftr = Fitter.auto(toas, model1, downhill=False)
+    ftr.fit_toas(maxiter=5)
+    aics = [akaike_information_criterion(model1, toas)]
+    model1 = ftr.model
+
+    T_span = toas.get_mjds().max() - toas.get_mjds().min()
+    setup_component = wavex_setup if component == "WaveX" else dmwavex_setup
+    setup_component(model1, T_span, n_freqs=1, freeze_params=False)
+
+    for _ in range(nharms_max):
+        ftr = Fitter.auto(toas, model1, downhill=False)
+        ftr.fit_toas(maxiter=5)
+        aics.append(akaike_information_criterion(ftr.model, toas))
+
+        model1 = ftr.model
+        if component == "WaveX":
+            model1.components[component].add_wavex_component(
+                (len(model1.components[component].get_indices()) + 1) / T_span,
+                frozen=False,
+            )
+        else:
+            model1.components[component].add_dmwavex_component(
+                (len(model1.components[component].get_indices()) + 1) / T_span,
+                frozen=False,
+            )
+
+    return np.argmin(aics), np.array(aics) - np.min(aics)
```

Comment:
Any tests if there are NaN's in this?  Or if more than one value is at the minimum (probably indicative of an error)?

Reply:
Added check.

Reply:
> Or if more than one value is at the minimum (probably indicative of an error)

This scenario is not an error, although it is very unlikely. As I understand it, it just means that two Nharm values are equally good. In this case, `np.argmin` will return the earliest occurrence of the minimum, which corresponds to a lower Nharm value. I think this is the correct behavior.

Reply:
My point is that if they are close but not equal, that's OK.  but if they are exactly equal then they are probably 0 or something like that, so potentially an error.

Reply:
I don't think that will lead to an error. All that we do with AIC values is that we take their minimum (using argmin) to find the best model. We don't do any other calculation with it. As long as it is a finite value (i.e., not nan or inf), it should be fine even if two values happen to be equal.

Reply:
But if they are exactly equal, it seems unlikely that this happens with good data values.  But if you don't think that's an issue, then I won't worry.

Reply:
wouldn't it also return an array of indices?  I think that would break your test, although maybe it's not used elsewhere

Reply:
`np.argmin` returns the first occurrance only.

For example, `np.argmin([2,4,5,1,2,1])` returns 3.